source/src/Internal.cc

```diff
@@ -28,229 +28,694 @@
-        throw core::StatusCodeException(STATUS_CODE_FAILURE);
+
+      dqm_long cpu_ticks1[4] = {0L};
+      dqm_long cpu_ticks2[4] = {0L};
+      readLinuxCpu(cpu_ticks1);
+      ::sleep(sampleTime);
+      readLinuxCpu(cpu_ticks2);
+
+      dqm_long userticks = (cpu_ticks2[0] + cpu_ticks2[3]) - (cpu_ticks1[0] + cpu_ticks1[3]);
+      dqm_long systicks = cpu_ticks2[1] - cpu_ticks1[1];
+      dqm_long idleticks = cpu_ticks2[2] - cpu_ticks1[2];
+      if (userticks < 0)
+        userticks = 0;
+      if (systicks < 0)
+        systicks = 0;
+      if (idleticks < 0)
+        idleticks = 0;
+      dqm_long totalticks = userticks + systicks + idleticks;
+      if (totalticks) {
+        stats.user = ((dqm_float)(100 * userticks)) / ((dqm_float)totalticks);
+        stats.sys = ((dqm_float)(100 * systicks)) / ((dqm_float)totalticks);
+        stats.tot = stats.user + stats.sys;
+        stats.idle = ((dqm_float)(100 * idleticks)) / ((dqm_float)totalticks);
       }
-      double wired = vmStat.wire_count; // In use, cannot go inactive
-      double active = vmStat.active_count; // Currently in use, can go inactive if not accessed for some time
-      double inactive = vmStat.inactive_count; // Was in use not long ago, still retrievable -> Not Sure wheter to include in used ram
-      double free = vmStat.free_count; // Real free ram usable instantly -> Always around 0-50mb max on my mac
-      double total = wired + active + inactive;
-
-      stats.rssused = (total - free) * pageSize;
-      stats.vmused = stats.rssused + swapStats.xsu_used;
-      stats.rssused /= unit;
-      stats.vmused /= unit;
-
-
-      // Process virtual and physical memory size
-      // Retrieved in Bytes
-      // Physical memory in this case is the actual memory taken by the process
-      // Virtual memory: Most processes are assigned ~4Go of virtual mem whether they use it or not. No idea what is governing this, or the real meaning of it.
-      struct task_basic_info_64 taskInfo;
-      mach_msg_type_number_t taskInfoCount = TASK_BASIC_INFO_64_COUNT;
-      if (KERN_SUCCESS != task_info(mach_task_self(), TASK_BASIC_INFO_64, (task_info_t) &taskInfo, &taskInfoCount))
-      {
-        dqm_error("[{0}] - Failed to get process memory statistics.", __FUNCTION__);
+    }
+
+    //-------------------------------------------------------------------------------------------------
+    /// Get VM stat for Linux.
+
+    static void linuxMemStats(MemoryStats &stats) {
+      TString s;
+      FILE *f = fopen("/proc/meminfo", "r");
+      if (!f) {
+        dqm_error("[{0}] - Failed to open '/proc/meminfo'", __FUNCTION__);
         throw core::StatusCodeException(STATUS_CODE_FAILURE);
       }
-      stats.rssproc = taskInfo.resident_size;
-      stats.rssproc /= unit;
-      stats.vmproc = taskInfo.resident_size; // Display physical size, as virtual_size is meaningless in this context
-      stats.vmproc /= unit;
+      dqm_long_long free = {0LL};
```

Comment:
Please change the name `free` here as it is reserved for memory

Reply:
Indeed, fixed now.